kube-runtime/src/utils/stream_subscribe.rs

```diff
@@ -0,0 +1,236 @@
+use core::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use futures::{stream, Stream};
+use pin_project::pin_project;
+use std::{fmt, sync::Arc};
+use tokio::sync::{broadcast, broadcast::error::RecvError};
+
+const CHANNEL_CAPACITY: usize = 128;
+
+/// Exposes the [`StreamSubscribe::subscribe()`] method which allows additional
+/// consumers of events from a stream without consuming the stream itself.
+///
+/// If a subscriber begins to lag behind the stream, it will receive an [`Error::Lagged`]
+/// error. The subscriber can then decide to abort its task or tolerate the lost events.
+///
+/// If the [`Stream`] is dropped or ends, any [`StreamSubscribe::subscribe()`] streams
+/// will also end.
+#[pin_project]
+#[must_use = "subscribers will not get events unless this stream is polled"]
+pub struct StreamSubscribe<S>
+where
+    S: Stream,
+{
+    #[pin]
+    stream: S,
+    sender: broadcast::Sender<Option<Arc<S::Item>>>,
+}
+
+impl<S: Stream> StreamSubscribe<S> {
+    pub fn new(stream: S) -> Self {
+        let (sender, _) = broadcast::channel(CHANNEL_CAPACITY);
+
+        Self { stream, sender }
+    }
+
+    /// Subscribe to events from this stream
+    #[must_use = "streams do nothing unless polled"]
+    pub fn subscribe(&self) -> impl Stream<Item = Result<Arc<S::Item>, Error>> {
+        stream::unfold(self.sender.subscribe(), |mut rx| async {
+            match rx.recv().await {
+                Ok(Some(obj)) => Some((Ok(obj), rx)),
+                Err(RecvError::Lagged(amt)) => Some((Err(Error::Lagged(amt)), rx)),
+                _ => None,
```

Comment:
Oh, we need to handle errors better here i think;
- if the watcher emits an error here, it will restart, but we will just close the stream

so either we need to return a watcher::Error and union in a Lagged error into that enum, or we need to have our error here be a union of our new error and watcher::Error

Reply:
ah, wait, `watcher` catches all of its errors though and retries.. maybe the `watcher` interface is a bit misleading here..